File: wails_backend/internal/services/appai.go
package services

import (
	"context"
	"fmt"
	"strings"

	"wails_backend/internal/db"
	"wails_backend/internal/lib"
	"wails_backend/internal/types"

	"github.com/firebase/genkit/go/ai"
)

const appAIAgentSystem = `You are a configuration assistant for Workspace Manager.
You edit ONLY the currently selected config set of the current app.

Use tools to inspect and change this set:
- env vars: list_vars, get_var, update_var, delete_var
- templates: list_templates, get_template, update_template (Handlebars {{VAR_NAME}} for env placeholders)
- run config: get_run_config, update_run_config
- project files: search_files (glob; gitignored omitted), read_file (relative path)

You must NOT:
- create, rename, or delete config sets
- invent template file paths; only update templates list_templates returns
- reformat template files unless the user asked

When updating run commands, send the full command list.
Edits are staged for the user to review. Reply in short markdown (lists, inline code). No JSON. No headings.`

type emptyIn struct{}

type keyIn struct {
	Key string `json:"key"`
}

type updateVarIn struct {
	Key   string `json:"key"`
	Value string `json:"value"`
}

type pathIn struct {
	FilePath string `json:"file_path"`
}

type updateTemplateIn struct {
	FilePath string `json:"file_path"`
	Content  string `json:"content"`
}

type runCmdIn struct {
	Label   *string `json:"label"`
	Command string  `json:"command"`
}

type updateRunIn struct {
	Mode     *string    `json:"mode,omitempty"`
	Commands []runCmdIn `json:"commands,omitempty"`
}

type searchIn struct {
	Pattern string `json:"pattern"`
}

type filePathIn struct {
	Path string `json:"path"`
}

type envPair struct {
	Key   string `json:"key"`
	Value string `json:"value"`
}

type runCmd struct {
	Label   *string `json:"label"`
	Command string  `json:"command"`
}

type appAIState struct {
	projectPath string
	env         map[string]string
	origEnv     map[string]string
	deletedEnv  map[string]bool
	templates   map[string]string
	origTmpl    map[string]string
	runMode     string
	runCommands []runCmd
	runDirty    bool
	calls       []types.AppAIToolCall
	emit        func(types.AppAIStreamEvent)
}

func (s *appAIState) record(name string, in, out any) any {
	call := types.AppAIToolCall{Name: name, Input: in, Output: out}
	s.calls = append(s.calls, call)
	if s.emit != nil {
		c := call
		s.emit(types.AppAIStreamEvent{Type: "tool", Call: &c})
	}
	return out
}

func newAppAIState(detail types.ConfigSetDetail, projectPath string) *appAIState {
	s := &appAIState{
		projectPath: projectPath,
		env:         map[string]string{},
		origEnv:     map[string]string{},
		deletedEnv:  map[string]bool{},
		templates:   map[string]string{},
		origTmpl:    map[string]string{},
		runMode:     "parallel",
	}
	for _, v := range detail.EnvVars {
		s.env[v.Key] = v.Value
		s.origEnv[v.Key] = v.Value
	}
	for _, t := range detail.Templates {
		s.templates[t.FilePath] = t.Content
		s.origTmpl[t.FilePath] = t.Content
	}
	if detail.RunConfig != nil {
		s.runMode = detail.RunConfig.Mode
		for _, c := range detail.RunConfig.Commands {
			label := c.Label
			s.runCommands = append(s.runCommands, runCmd{Label: label, Command: c.Command})
		}
	}
	return s
}

func (s *appAIState) patch() types.AppAIPatch {
	p := types.AppAIPatch{}
	var upsert []types.AppAIEnvUpsert
	for k, v := range s.env {
		orig, ok := s.origEnv[k]
		if !ok || orig != v {
			upsert = append(upsert, types.AppAIEnvUpsert{Key: k, Value: v})
		}
	}
	var del []string
	for k := range s.deletedEnv {
		del = append(del, k)
	}
	if len(upsert) > 0 || len(del) > 0 {
		p.Env = &types.AppAIEnvPatch{Upsert: upsert, Delete: del}
	}
	for path, content := range s.templates {
		orig, ok := s.origTmpl[path]
		if !ok || orig != content {
			p.Templates = append(p.Templates, types.AppAITemplatePatch{FilePath: path, Content: content})
		}
	}
	if s.runDirty {
		cmds := make([]types.AppAIRunCommand, 0, len(s.runCommands))
		for _, c := range s.runCommands {
			cmds = append(cmds, types.AppAIRunCommand{Label: c.Label, Command: c.Command})
		}
		mode := s.runMode
		p.Run = &types.AppAIRunPatch{Mode: &mode, Commands: cmds}
	}
	return p
}

func (s *appAIState) listVars() map[string]any {
	vars := make([]envPair, 0, len(s.env))
	for k, v := range s.env {
		vars = append(vars, envPair{Key: k, Value: v})
	}
	return map[string]any{"vars": vars}
}

func (s *appAIState) getVar(key string) map[string]any {
	k := strings.TrimSpace(key)
	if k == "" {
		return map[string]any{"error": "key is required"}
	}
	v, ok := s.env[k]
	if !ok {
		return map[string]any{"error": "env var not found: " + k}
	}
	return map[string]any{"key": k, "value": v}
}

func (s *appAIState) updateVar(key, value string) map[string]any {
	k := strings.TrimSpace(key)
	if k == "" {
		return map[string]any{"error": "key is required"}
	}
	delete(s.deletedEnv, k)
	s.env[k] = value
	return map[string]any{"ok": true, "key": k, "value": value}
}

func (s *appAIState) deleteVar(key string) map[string]any {
	k := strings.TrimSpace(key)
	if k == "" {
		return map[string]any{"error": "key is required"}
	}
	_, live := s.env[k]
	_, orig := s.origEnv[k]
	if !live && !orig {
		return map[string]any{"error": "env var not found: " + k}
	}
	delete(s.env, k)
	if orig {
		s.deletedEnv[k] = true
	}
	return map[string]any{"ok": true, "key": k}
}

func (s *appAIState) listTemplates() map[string]any {
	paths := make([]string, 0, len(s.templates))
	for p := range s.templates {
		paths = append(paths, p)
	}
	return map[string]any{"file_paths": paths}
}

func (s *appAIState) getTemplate(filePath string) map[string]any {
	p := strings.ReplaceAll(strings.TrimSpace(filePath), "\\", "/")
	if p == "" {
		return map[string]any{"error": "file_path is required"}
	}
	c, ok := s.templates[p]
	if !ok {
		return map[string]any{"error": "template not found: " + p}
	}
	return map[string]any{"file_path": p, "content": c}
}

func (s *appAIState) updateTemplate(filePath, content string) map[string]any {
	p := strings.ReplaceAll(strings.TrimSpace(filePath), "\\", "/")
	if p == "" {
		return map[string]any{"error": "file_path is required"}
	}
	if _, ok := s.templates[p]; !ok {
		return map[string]any{"error": "template not on this config set: " + p}
	}
	s.templates[p] = content
	return map[string]any{"ok": true, "file_path": p}
}

func (s *appAIState) getRun() map[string]any {
	return map[string]any{"mode": s.runMode, "commands": s.runCommands}
}

func (s *appAIState) updateRun(in updateRunIn) map[string]any {
	if in.Mode != nil && *in.Mode != "" {
		m := *in.Mode
		if m != "parallel" && m != "sequential" {
			return map[string]any{"error": `mode must be "parallel" or "sequential"`}
		}
		s.runMode = m
		s.runDirty = true
	}
	if in.Commands != nil {
		cmds := make([]runCmd, 0, len(in.Commands))
		for _, c := range in.Commands {
			command := strings.TrimSpace(c.Command)
			if command == "" {
				continue
			}
			var label *string
			if c.Label != nil {
				l := strings.TrimSpace(*c.Label)
				if l != "" {
					label = &l
				}
			}
			cmds = append(cmds, runCmd{Label: label, Command: command})
		}
		s.runCommands = cmds
		s.runDirty = true
	}
	if !s.runDirty {
		return map[string]any{"error": "provide mode and/or commands"}
	}
	return map[string]any{"ok": true, "mode": s.runMode, "commands": s.runCommands}
}

func (s *appAIState) searchFiles(pattern string) map[string]any {
	files, truncated, errMsg := lib.SearchProjectFiles(s.projectPath, pattern)
	if errMsg != "" {
		return map[string]any{"error": errMsg}
	}
	return map[string]any{"files": files, "truncated": truncated}
}

func (s *appAIState) readFile(path string) map[string]any {
	content, rel, errMsg := lib.ReadProjectFile(s.projectPath, path)
	if errMsg != "" {
		return map[string]any{"error": errMsg}
	}
	return map[string]any{"file_path": rel, "content": content}
}

func (s *appAIState) tools() []ai.ToolRef {
	return []ai.ToolRef{
		ai.NewTool("list_vars", "List env var keys and values on the active config set.",
			func(_ *ai.ToolContext, in emptyIn) (any, error) {
				return s.record("list_vars", in, s.listVars()), nil
			}),
		ai.NewTool("get_var", "Get one env var by key.",
			func(_ *ai.ToolContext, in keyIn) (any, error) {
				return s.record("get_var", in, s.getVar(in.Key)), nil
			}),
		ai.NewTool("update_var", "Create or update an env var on the active config set.",
			func(_ *ai.ToolContext, in updateVarIn) (any, error) {
				return s.record("update_var", in, s.updateVar(in.Key, in.Value)), nil
			}),
		ai.NewTool("delete_var", "Delete an env var from the active config set.",
			func(_ *ai.ToolContext, in keyIn) (any, error) {
				return s.record("delete_var", in, s.deleteVar(in.Key)), nil
			}),
		ai.NewTool("list_templates", "List template file paths on the active config set (no content).",
			func(_ *ai.ToolContext, in emptyIn) (any, error) {
				return s.record("list_templates", in, s.listTemplates()), nil
			}),
		ai.NewTool("get_template", "Get the full content of one template by file_path.",
			func(_ *ai.ToolContext, in pathIn) (any, error) {
				return s.record("get_template", in, s.getTemplate(in.FilePath)), nil
			}),
		ai.NewTool("update_template", "Replace the content of an existing template. file_path must already be on this config set.",
			func(_ *ai.ToolContext, in updateTemplateIn) (any, error) {
				return s.record("update_template", in, s.updateTemplate(in.FilePath, in.Content)), nil
			}),
		ai.NewTool("get_run_config", "Get run mode and commands for the active config set.",
			func(_ *ai.ToolContext, in emptyIn) (any, error) {
				return s.record("get_run_config", in, s.getRun()), nil
			}),
		ai.NewTool("update_run_config", "Update run mode and/or replace the full command list for this config set.",
			func(_ *ai.ToolContext, in updateRunIn) (any, error) {
				return s.record("update_run_config", in, s.updateRun(in)), nil
			}),
		ai.NewTool("search_files", "Glob search under the app project directory. Respects .gitignore. Returns matching relative paths.",
			func(_ *ai.ToolContext, in searchIn) (any, error) {
				return s.record("search_files", in, s.searchFiles(in.Pattern)), nil
			}),
		ai.NewTool("read_file", "Read a text file under the app project directory. Path must be relative and stay inside the project.",
			func(_ *ai.ToolContext, in filePathIn) (any, error) {
				return s.record("read_file", in, s.readFile(in.Path)), nil
			}),
	}
}

func buildAppAIAgentPrompt(appName, projectPath, setName string, setID int64, history []types.AppAIChatTurn, instruction string) string {
	var b strings.Builder
	fmt.Fprintf(&b, "App: %s\nProject path: %s\n\nActive config set (ONLY edit this one):\nid: %d\nname: %s\n\n", appName, projectPath, setID, setName)
	b.WriteString("Use tools to read or edit env vars, templates, and run config. Use search_files and read_file when you need project files.\n\nPrior conversation:\n")
	if len(history) == 0 {
		b.WriteString("(none)\n")
	} else {
		for _, t := range history {
			who := "Assistant"
			if t.Role == "user" {
				who = "User"
			}
			fmt.Fprintf(&b, "%s: %s\n", who, t.Text)
		}
	}
	fmt.Fprintf(&b, "\nUser instruction:\n%s", strings.TrimSpace(instruction))
	return b.String()
}

func patchHasEdits(p types.AppAIPatch) bool {
	if p.Env != nil && (len(p.Env.Upsert) > 0 || len(p.Env.Delete) > 0) {
		return true
	}
	if len(p.Templates) > 0 {
		return true
	}
	return p.Run != nil
}

// AppChatAI runs the config-set agent with tools. Updates are staged in the
// returned patch for the UI to review; nothing is written until the user applies.
func AppChatAI(ctx context.Context, d *db.DB, appID, setID int64, history []types.AppAIChatTurn, instruction string, emit func(types.AppAIStreamEvent)) (types.AppAIChatResult, error) {
	instruction = strings.TrimSpace(instruction)
	if instruction == "" {
		return types.AppAIChatResult{}, fmt.Errorf("instruction is required")
	}
	app, err := d.GetAppT(ctx, appID)
	if err != nil {
		return types.AppAIChatResult{}, err
	}
	set, err := d.GetConfigSetT(ctx, setID)
	if err != nil {
		return types.AppAIChatResult{}, err
	}
	if set.AppID != app.ID {
		return types.AppAIChatResult{}, fmt.Errorf("Config set not found")
	}
	detail, err := loadConfigSetDetail(ctx, d, set)
	if err != nil {
		return types.AppAIChatResult{}, err
	}
	state := newAppAIState(detail, app.ProjectPath)
	state.emit = emit

	store, err := LoadAIStore()
	if err != nil {
		return types.AppAIChatResult{}, err
	}
	_, conn, err := store.ActiveAIConnection()
	if err != nil {
		return types.AppAIChatResult{}, err
	}

	prompt := buildAppAIAgentPrompt(app.Name, app.ProjectPath, set.Name, set.ID, history, instruction)
	var onText func(string)
	if emit != nil {
		onText = func(t string) {
			emit(types.AppAIStreamEvent{Type: "text", Text: t})
		}
	}
	text, err := runGeneration(ctx, conn.Provider, conn, appAIAgentSystem, prompt, state.tools(), onText)
	if err != nil {
		return types.AppAIChatResult{}, err
	}
	text = strings.TrimSpace(text)
	if text == "" {
		text = "Done."
	}
	patch := state.patch()
	patch.Message = text
	if !patchHasEdits(patch) {
		return types.AppAIChatResult{
			Text:      text,
			Patch:     types.AppAIPatch{Message: text},
			ToolCalls: state.calls,
		}, nil
	}
	return types.AppAIChatResult{Text: text, Patch: patch, ToolCalls: state.calls}, nil
}

func loadConfigSetDetail(ctx context.Context, d *db.DB, set types.ConfigSet) (types.ConfigSetDetail, error) {
	envVars, err := d.ListEnvVarsT(ctx, set.ID)
	if err != nil {
		return types.ConfigSetDetail{}, err
	}
	templates, err := d.ListTemplatesT(ctx, set.ID)
	if err != nil {
		return types.ConfigSetDetail{}, err
	}
	runCfg, err := d.GetRunConfigByConfigSetT(ctx, set.ID)
	if err != nil {
		return types.ConfigSetDetail{}, err
	}
	return types.ConfigSetDetail{ConfigSet: set, EnvVars: envVars, Templates: templates, RunConfig: runCfg}, nil
}

File: frontend/lib/app-ai.ts
import type { ConfigSetDetail, RunMode } from "./types"

export const APP_AI_SYSTEM_PROMPT = `You are a configuration assistant for Workspace Manager.
You edit ONLY the currently selected config set of the current app.

Use tools to inspect and change this set. Do not expect env vars, templates, or run config in the user message.

You must NOT:
- create, rename, delete, or edit any other config set
- invent template file paths
- reformat template files unless the user asked

Reply in short markdown (lists, inline code). No JSON. No headings. Edits are applied through tools.`

export type AppAIChatTurn = {
  role: "user" | "assistant"
  text: string
  /** Prior tool calls for this assistant turn (replayed as Genkit tool messages). */
  tools?: AppAIToolCall[]
}

export type AppAIToolCall = {
  name: string
  input: unknown
  output: unknown
}

export type AppAIStreamEvent =
  | { type: "tool"; call: AppAIToolCall }
  | { type: "text"; text: string }
  | {
      type: "done"
      text: string
      patch: AppAIPatch
      toolCalls?: AppAIToolCall[]
    }
  | { type: "error"; error: string }

export type AppAIChatResult = {
  text: string
  patch: AppAIPatch
  toolCalls?: AppAIToolCall[]
}

export function summarizeAIToolInput(input: unknown, max = 72): string {
  let s: string
  try {
    s = JSON.stringify(input) ?? ""
  } catch {
    s = String(input)
  }
  if (s === "{}" || s === "null" || s === "") return ""
  return s.length <= max ? s : `${s.slice(0, max - 1)}…`
}

export type AppAIEnvUpsert = { key: string; value: string }

export type AppAIPatch = {
  message: string
  env?: {
    upsert?: AppAIEnvUpsert[]
    delete?: string[]
  }
  templates?: { file_path: string; content: string }[]
  run?: {
    mode?: RunMode
    commands?: { label?: string | null; command: string }[]
  }
}

export function buildAppAIPrompt({
  appName,
  projectPath,
  configSet,
  instruction,
}: {
  appName: string
  projectPath: string
  configSet: ConfigSetDetail
  /** @deprecated prior turns are Genkit messages, not prompt text */
  history?: AppAIChatTurn[]
  instruction: string
  /** @deprecated templates are fetched via tools */
  templatePaths?: string[]
}): string {
  return `App: ${appName}
Project path: ${projectPath}

Active config set (ONLY edit this one):
id: ${configSet.id}
name: ${configSet.name}

Use tools to read or edit env vars, templates, and run config. Use search_files and read_file when you need project files.

User instruction:
${instruction.trim()}`
}

export function stripAIFences(raw: string): string {
  let text = raw.trim()
  const codeBlockMatch = text.match(
    /^```(?:[a-zA-Z0-9_.-]+)?\r?\n([\s\S]*?)\r?\n```$/
  )
  if (codeBlockMatch && codeBlockMatch[1] !== undefined) {
    return codeBlockMatch[1]
  }
  if (text.startsWith("```") && text.endsWith("```")) {
    text = text.slice(3, -3).trim()
    const firstNewline = text.indexOf("\n")
    if (firstNewline !== -1) {
      const firstLine = text.slice(0, firstNewline).trim()
      if (/^[a-zA-Z0-9_.-]+$/.test(firstLine)) {
        text = text.slice(firstNewline + 1)
      }
    }
    return text
  }
  return text
}

function asString(v: unknown): string | undefined {
  return typeof v === "string" ? v : undefined
}

function parseEnv(raw: unknown): AppAIPatch["env"] | undefined {
  if (!raw || typeof raw !== "object") return undefined
  const obj = raw as Record<string, unknown>
  const upsert: AppAIEnvUpsert[] = []
  if (Array.isArray(obj.upsert)) {
    for (const item of obj.upsert) {
      if (!item || typeof item !== "object") continue
      const rec = item as Record<string, unknown>
      const key = asString(rec.key)?.trim()
      if (!key) continue
      upsert.push({ key, value: asString(rec.value) ?? "" })
    }
  }
  const del: string[] = []
  if (Array.isArray(obj.delete)) {
    for (const item of obj.delete) {
      if (typeof item === "string" && item.trim()) del.push(item.trim())
    }
  }
  if (upsert.length === 0 && del.length === 0) return undefined
  return {
    upsert: upsert.length ? upsert : undefined,
    delete: del.length ? del : undefined,
  }
}

function parseTemplates(raw: unknown): AppAIPatch["templates"] | undefined {
  if (!Array.isArray(raw)) return undefined
  const out: { file_path: string; content: string }[] = []
  for (const item of raw) {
    if (!item || typeof item !== "object") continue
    const rec = item as Record<string, unknown>
    const file_path = asString(rec.file_path)?.trim()
    const content = asString(rec.content)
    if (!file_path || content == null) continue
    out.push({ file_path, content })
  }
  return out.length ? out : undefined
}

function parseRun(raw: unknown): AppAIPatch["run"] | undefined {
  if (!raw || typeof raw !== "object") return undefined
  const obj = raw as Record<string, unknown>
  const modeRaw = asString(obj.mode)
  const mode: RunMode | undefined =
    modeRaw === "parallel" || modeRaw === "sequential" ? modeRaw : undefined
  const commands: { label?: string | null; command: string }[] = []
  if (Array.isArray(obj.commands)) {
    for (const item of obj.commands) {
      if (!item || typeof item !== "object") continue
      const rec = item as Record<string, unknown>
      const command = asString(rec.command)?.trim()
      if (!command) continue
      const label = asString(rec.label)?.trim() || null
      commands.push({ label, command })
    }
  }
  if (!mode && commands.length === 0) return undefined
  return {
    mode,
    commands: commands.length ? commands : undefined,
  }
}

export function parseAppAIResponse(raw: string): AppAIPatch {
  const text = stripAIFences(raw)
  const start = text.indexOf("{")
  const end = text.lastIndexOf("}")
  if (start === -1 || end <= start) {
    return { message: text || "Done." }
  }
  try {
    const obj = JSON.parse(text.slice(start, end + 1)) as Record<string, unknown>
    const message = asString(obj.message)?.trim() || "Done."
    return {
      message,
      env: parseEnv(obj.env),
      templates: parseTemplates(obj.templates),
      run: parseRun(obj.run),
    }
  } catch {
    return { message: text || "Done." }
  }
}

export function patchHasEdits(patch: AppAIPatch): boolean {
  return !!(
    patch.env?.upsert?.length ||
    patch.env?.delete?.length ||
    patch.templates?.length ||
    patch.run
  )
}
